File: istaroth/agd/renderables/creature.py
"""Creature (living-beings archive) processing and rendering."""

import hashlib

from istaroth import utils
from istaroth.agd import (
    agd_types,
    id_types,
    issues,
    localization,
    processed_types,
    repo,
)
from istaroth.text import types as text_types


def _get_unique_monster_special_name_hash(
    special_names: agd_types.MonsterSpecialNameExcelConfigData,
    special_name_lab_id: id_types.MonsterSpecialNameLabId,
) -> id_types.TextMapHash | None:
    matches = [
        entry
        for entry in special_names
        if entry["specialNameLabID"] == special_name_lab_id
    ]
    if not matches:
        raise ValueError(f"Missing monster special-name lab ID {special_name_lab_id}")
    return matches[0]["specialNameTextMapHash"] if len(matches) == 1 else None


def _get_creature_info(
    codex_id: id_types.AnimalCodexId, *, data_repo: repo.DataRepo
) -> processed_types.CreatureInfo:
    """Get a living-beings archive entry (monster or wildlife) by its codex id."""
    entry = data_repo.load_animal_codex_excel_config_data()[codex_id]
    text_map = data_repo.build_text_map_tracker()

    if (description := text_map.get_optional(entry["descTextMapHash"])) is None:
        raise ValueError(f"Missing description for creature codex {codex_id}")

    describe_id = entry["describeId"]
    title: str | None = None
    special_name: str | None = None
    match entry["type"]:
        case "CODEX_MONSTER":
            describe = data_repo.load_monster_describe_excel_config_data()[describe_id]
            title = text_map.get_optional(
                data_repo.load_monster_title_excel_config_data()[describe["titleID"]][
                    "titleNameTextMapHash"
                ]
            )
            special_name_hash = _get_unique_monster_special_name_hash(
                data_repo.load_monster_special_name_excel_config_data(),
                describe["specialNameLabID"],
            )
            special_name = (
                None
                if special_name_hash is None
                else text_map.get_optional(special_name_hash)
            )
            name_hash = describe["nameTextMapHash"]
        case "CODEX_ANIMAL":
            name_hash = data_repo.load_animal_describe_excel_config_data()[describe_id][
                "nameTextMapHash"
            ]
        case other:
            raise ValueError(f"Unknown codex type {other!r} for creature {codex_id}")

    if (name := text_map.get_optional(name_hash)) is None:
        raise ValueError(f"Missing name for creature codex {codex_id}")

    return processed_types.CreatureInfo(
        codex_id=codex_id,
        name=name,
        special_name=special_name if special_name != name else None,
        title=title if title not in (name, special_name) else None,
        description=description,
    )


def get_creature_group_info(
    subtype: str, *, data_repo: repo.DataRepo
) -> processed_types.CreatureGroupInfo:
    """Get all creatures in one codex subType group, in archive order."""
    entries = sorted(
        (
            entry
            for entry in data_repo.load_animal_codex_excel_config_data().values()
            if entry["subType"] == subtype and not entry["isDisuse"]
        ),
        key=lambda entry: (entry["sortOrder"], entry["id"]),
    )
    if not entries:
        raise ValueError(f"No creatures found for codex subType {subtype!r}")

    return processed_types.CreatureGroupInfo(
        subtype=subtype,
        type_label=localization.get_creature_type_label(
            entries[0]["type"], language=data_repo.language
        ),
        subtype_label=localization.get_creature_subtype_label(
            subtype, language=data_repo.language
        ),
        creatures=[
            _get_creature_info(entry["id"], data_repo=data_repo) for entry in entries
        ],
    )


def render_creature_group(
    group_info: processed_types.CreatureGroupInfo,
) -> processed_types.RenderedItem:
    """Render one codex subType group of creatures into a single RAG-suitable file."""
    group_id = int(
        hashlib.sha256(group_info.subtype.encode("utf-8")).hexdigest()[:12], base=16
    )
    filename = f"{group_id}_{group_info.subtype}.txt"
    title = group_info.subtype_label

    content_lines = [f"# {title} ({group_info.type_label})\n"]
    for creature in group_info.creatures:
        content_lines.append(f"## {creature.name}")
        if creature.special_name is not None:
            content_lines.append(creature.special_name)
        if creature.title is not None:
            content_lines.append(f"Also known as: {creature.title}")
        content_lines.append("")
        content_lines.append(creature.description)
        content_lines.append("")

    return processed_types.RenderedItem(
        text_metadata=text_types.TextMetadata(
            category=text_types.TextCategory.AGD_CREATURE,
            title=title,
            id=group_id,
            relative_path=f"{text_types.TextCategory.AGD_CREATURE.value}/{filename}",
        ),
        content="\n".join(content_lines).rstrip(),
    )

File: rust/istaroth-agd-regen/src/renderables/creature.rs
//! Port of renderables/creature.py: living-beings archive (monsters and
//! wildlife) grouped by codex subType.

use crate::firstseen::Domain;
use crate::issues::Scope;
use crate::lang::Language;
use crate::rendered_item::RenderedItem;
use crate::repo::Repo;
use crate::util;
use crate::vh::ValueExt;
use anyhow::{Result, anyhow, bail};
use rustc_hash::FxHashSet;
use serde_json::Value;

/// (CHS, non-CHS) per-pass error limits (see e.g. `artifact::ERROR_LIMITS`).
pub const ERROR_LIMITS: (usize, usize) = (0, 2);

const CREATURE_TYPE_LABELS: [(&str, &str, &str); 2] = [
    ("CODEX_MONSTER", "魔物", "Monsters"),
    ("CODEX_ANIMAL", "野生生物", "Wildlife"),
];
const CREATURE_SUBTYPE_LABELS: [(&str, &str, &str); 12] = [
    ("CODEX_SUBTYPE_ELEMENTAL", "元素生命", "Elemental Lifeforms"),
    ("CODEX_SUBTYPE_HILICHURL", "丘丘部族", "Hilichurls"),
    ("CODEX_SUBTYPE_ABYSS", "深渊", "The Abyss"),
    ("CODEX_SUBTYPE_FATUI", "愚人众", "Fatui"),
    ("CODEX_SUBTYPE_AUTOMATRON", "自律机关", "Automatons"),
    ("CODEX_SUBTYPE_HUMAN", "人类势力", "Human Factions"),
    ("CODEX_SUBTYPE_BEAST", "异种魔兽", "Mystical Beasts"),
    ("CODEX_SUBTYPE_BOSS", "首领", "Bosses"),
    ("CODEX_SUBTYPE_ANIMAL", "走兽", "Beasts"),
    ("CODEX_SUBTYPE_AVIARY", "飞禽", "Avian Kind"),
    ("CODEX_SUBTYPE_FISH", "鱼类", "Fish"),
    ("CODEX_SUBTYPE_CRITTER", "小型生物", "Critters"),
];

fn label_of(
    table: &[(&str, &'static str, &'static str)],
    key: &str,
    language: Language,
) -> Result<&'static str> {
    table
        .iter()
        .find(|(k, _, _)| *k == key)
        .map(|(_, chs, eng)| match language {
            Language::Chs => *chs,
            Language::Eng => *eng,
        })
        .ok_or_else(|| anyhow!("unknown codex label {key}"))
}

struct CreatureInfo {
    codex_id: i64,
    name: String,
    special_name: Option<String>,
    title: Option<String>,
    description: String,
}

fn creature_info(repo: &Repo, scope: &Scope, codex_id: i64) -> Result<CreatureInfo> {
    let entry = repo
        .excel
        .animal_codex
        .get(&codex_id)
        .ok_or_else(|| anyhow!("unknown animal codex {codex_id}"))?;
    let description = repo.tm.get_required(entry.i("descTextMapHash")?, scope)?;
    let describe_id = entry.i("describeId")?;

    let mut title: Option<String> = None;
    let mut special_name: Option<String> = None;
    let name_hash: i64;
    match entry.s("type")? {
        "CODEX_MONSTER" => {
            let describe = repo
                .excel
                .monster_describe
                .get(&describe_id)
                .ok_or_else(|| anyhow!("unknown monster describe {describe_id}"))?;
            let title_id = describe.i("titleID")?;
            let title_entry = repo
                .excel
                .monster_title
                .get(&title_id)
                .ok_or_else(|| anyhow!("unknown monster title {title_id}"))?;
            title = repo
                .tm
                .get_optional(title_entry.i("titleNameTextMapHash")?, scope)?;
            let lab_id = describe.i("specialNameLabID")?;
            let mut matches: Vec<&Value> = Vec::new();
            for entry in &repo.excel.monster_special_name {
                if entry.i("specialNameLabID")? == lab_id {
                    matches.push(entry);
                }
            }
            if matches.is_empty() {
                bail!("Missing monster special-name lab ID {lab_id}");
            }
            if matches.len() == 1 {
                special_name = repo
                    .tm
                    .get_optional(matches[0].i("specialNameTextMapHash")?, scope)?;
            }
            name_hash = describe.i("nameTextMapHash")?;
        }
        "CODEX_ANIMAL" => {
            name_hash = repo
                .excel
                .animal_describe
                .get(&describe_id)
                .ok_or_else(|| anyhow!("unknown animal describe {describe_id}"))?
                .i("nameTextMapHash")?;
        }
        other => bail!("Unknown codex type {other:?} for creature {codex_id}"),
    }
    let name = repo.tm.get_required(name_hash, scope)?;
    let special_name = special_name.filter(|s| *s != name);
    let title = title
        .filter(|t| Some(t.as_str()) != Some(name.as_str()) && Some(t) != special_name.as_ref());
    Ok(CreatureInfo {
        codex_id,
        name,
        special_name,
        title,
        description,
    })
}

/// Creatures pass discovery: sorted distinct non-disused codex subTypes.
pub fn discover(repo: &Repo) -> Result<Vec<String>> {
    let mut subtypes: FxHashSet<String> = FxHashSet::default();
    for entry in repo.excel.animal_codex.values() {
        if !entry.b("isDisuse")? {
            subtypes.insert(entry.s("subType")?.to_string());
        }
    }
    let mut v: Vec<String> = subtypes.into_iter().collect();
    v.sort();
    Ok(v)
}

/// Creatures: one codex subType group per file.
pub fn process(repo: &Repo, scope: &Scope, subtype: &str) -> Result<Option<RenderedItem>> {
    let mut entries: Vec<(i64, i64, &Value)> = Vec::new();
    for entry in repo.excel.animal_codex.values() {
        if entry.s("subType")? == subtype && !entry.b("isDisuse")? {
            entries.push((entry.i("sortOrder")?, entry.i("id")?, entry));
        }
    }
    entries.sort_by_key(|(sort_order, id, _)| (*sort_order, *id));
    if entries.is_empty() {
        bail!("No creatures found for codex subType {subtype:?}");
    }
    let type_label = label_of(
        &CREATURE_TYPE_LABELS,
        entries[0].2.s("type")?,
        repo.language,
    )?;
    let subtype_label = label_of(&CREATURE_SUBTYPE_LABELS, subtype, repo.language)?;
    let creatures: Vec<CreatureInfo> = entries
        .iter()
        .map(|(_, id, _)| creature_info(repo, scope, *id))
        .collect::<Result<_>>()?;

    let group_id = util::sha256_id(subtype);
    let filename = format!("{group_id}_{subtype}.txt");
    let title = subtype_label;
    let mut content_lines = vec![format!("# {title} ({type_label})\n")];
    for creature in &creatures {
        content_lines.push(format!("## {}", creature.name));
        if let Some(special_name) = &creature.special_name {
            content_lines.push(special_name.clone());
        }
        if let Some(t) = &creature.title {
            content_lines.push(format!("Also known as: {t}"));
        }
        content_lines.push(String::new());
        content_lines.push(creature.description.clone());
        content_lines.push(String::new());
    }
    let versions = repo
        .first_seen
        .resolve_ints(Domain::AnimalCodex, creatures.iter().map(|c| c.codex_id))?;
    Ok(Some(RenderedItem::new(
        "agd_creature",
        title.to_string(),
        group_id,
        filename,
        versions,
        content_lines.join("\n").trim_end().to_string(),
    )))
}
